common/include/ICamera.hpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _I_CAMERA_HPP_
+#define _I_CAMERA_HPP_
+
+#include <nabla.h>
+#include <iostream>
+#include <cstdio>
+#include <fstream>
+#include <chrono>
+
+#include "camera/ICameraControl.hpp"
+
+namespace nbl::hlsl // TODO: DIFFERENT NAMESPACE
+{
+
+template<ProjectionMatrix T = float64_t4x4>
+class ICamera : public ICameraController<typename T>
+{ 
+public:
+	using base_t = typename ICameraController<typename T>;
+
+	struct Traits
+	{
+		using controller_t = base_t;
+		using projection_t = typename controller_t::projection_t;
+		using gimbal_t = typename controller_t::CGimbal;
+		using controller_virtual_event_t = typename controller_t::CVirtualEvent;
+	};
+
+	ICamera(core::smart_refctd_ptr<typename Traits::gimbal_t>&& gimbal, core::smart_refctd_ptr<typename Traits::projection_t> projection, const float32_t3& target = {0,0,0})
+		: base_t(core::smart_refctd_ptr(gimbal)), m_projection(core::smart_refctd_ptr(projection)) { setTarget(target); }
```

Comment:
not sure that we want gimbals to be freely allocatable on the heap and refcounted, I'm thinking that camera should OWN them and they should live inside the camera as a member (and not inherit from IReferenceCounted)